### PR TITLE
[WIP] Reimplementing np.mean with overload and adding axis parameter

### PR DESCRIPTION
**Pull request by luk-f-a**
Partially addresses #1269 and #4245 
Requires #4472 (first 3 commits come from that PR), #5060 and #5711, which fix intermediate problems.

### CODE REVIEW

**Comment by luk-f-a** [comment]
closing to open a new branch to avoid merge conflicts with #4472

File: numba/targets/arraymath.py [hunk 1]
```diff
@@ -399,22 +399,52 @@ def array_cumprod_impl(arr):
...
+        if isinstance(arr.dtype, (types.Integer, types.Boolean)):
+            ret_dtype = np.float64
+        elif isinstance(arr.dtype, (types.Float, types.Complex)):
+            ret_dtype = arr.dtype
+        else:
+            raise TypeError(("np.mean is not supported on {} arrays. "
+                            "It supports boolean, integer, float "
+                            "and complex arrays").format(arr.dtype))
+        # dispatch based on whether there's an axis parameter and its type
+        if axis is None:
```
Comment (sklam) [inline numba/targets/arraymath.py]:
```suggestion
        if axis is None or isinstance(axis, types.Omitted):
```
will get rid of the error.  It is part of a weird Numba behavior.  See https://github.com/numba/numba/issues/4694

**Comment by luk-f-a** [comment]
Status: The current `mean` is able to deal with np.datetime64 but an overload-based version fails because of #4540. It was agreed with @stuartarchibald that breaking support for np.datetime64 would be wrong, so this PR is on hold until that issue is resolved.

**Comment by luk-f-a** [comment]
current status: #5060 is almost done and will fix #4540 which was blocking this PR.
However, another issue came up: timedelta could not be divided by a number, which is needed to calculate the mean. So I fixed that in #5711. However, that's not the end of it, there's an additional issue: after changing the implementation from `lower_builtin` to `overload` the new function is not registered in `npydecl.registry`. This is a problem for `ir_utils.get_np_ufunc_typ`, which looks the function there, leading to a failure in the tests `test_parfor.test_var` and `test_parfor.test_std`

**Comment by N720720** [comment]
It looks like we are so close to get this working. How can we register the new function in npydecl.registry? How to contribute to this issue?

**Comment by sklam** [comment]
I pushed a merge main commit to resolve conflict so we get new CI run. There are changes to `get_np_ufunc_typ` that may change things.

**Pull request closed without merging**